### ISSUE TITLE
Could kernel help with restoring OS state on application exit?

### ISSUE THREAD

**Issue by mikrosk** [freemint/freemint#320]
This has been discussed a couple of times in the list (can't find it right now). The basic idea is that the kernel would maintain a list of safe / default values for various sensitive addresses: audio/video registers, vectors, etc. So when an application crashes (due to MP or just because it's nasty) it wont take the whole OS down with it.

In a further implementation this could have been fine-tuned in PRG flags, i.e. I know that this application is not very stable and/or doesn't restore video resolution properly, so I mark it as "refresh the video regs upon exit".

Of course, it's not totally straight-forward, e.g. the kernel would need to track/follow the XBRA protocol (maybe even have something like a private XBRA for its own needs for apps which do not support it) but overall I think this could be quite useful?

**Comment by th-otto** [comment]
I think that would be hard to implement. Imagine
- PRG1 starts, modifies some values
- PRG2 starts, modifies some values
- PRG1 crashes

Now kernel would restore state before PRG1 was started, but that was already modified by PRG2. Maybe causing PRG2 also to crash.

Also "nasty" programs usually don't use XBRA, so the kernel cannot correctly use XBRA to unlink their vectors.

Generally, in a multi-tasking environment its not safe to mess with vectors, no matter how "cleanly" that is done.

But you can maybe achieve something similar by writing a wrapper program that does those saves/restores. No need to modify the kernel for this.

**Comment by mikrosk** [comment]
Yeah, that is what I meant by the need of some kind of private tracking: the kernel would need to track which one was the first, second etc. This way the kernel would be way more useful because it is the kernel which manages scheduling.
